File: src/RequestKit.Shared/wwwroot/js/regexInterop.js
export function execute(pattern, testString, flags) {
    try {
        const regex = new RegExp(pattern, flags);
        const matches = [];

        if (flags.includes("g")) {
            let match;
            while ((match = regex.exec(testString)) !== null) {
                const groups = [];
                for (let i = 1; i < match.length; i++) {
                    groups.push({
                        groupIndex: i,
                        name: null,
                        value: match[i] ?? "",
                        index: match.index,
                        length: (match[i] ?? "").length
                    });
                }
                // Named groups
                if (match.groups) {
                    for (const [name, value] of Object.entries(match.groups)) {
                        const existing = groups.find(g => g.value === (value ?? "") && !g.name);
                        if (existing) existing.name = name;
                    }
                }
                matches.push({
                    index: match.index,
                    length: match[0].length,
                    value: match[0],
                    groups
                });
                if (!match[0].length) regex.lastIndex++;
            }
        } else {
            const match = regex.exec(testString);
            if (match) {
                const groups = [];
                for (let i = 1; i < match.length; i++) {
                    groups.push({
                        groupIndex: i,
                        name: null,
                        value: match[i] ?? "",
                        index: match.index,
                        length: (match[i] ?? "").length
                    });
                }
                if (match.groups) {
                    for (const [name, value] of Object.entries(match.groups)) {
                        const existing = groups.find(g => g.value === (value ?? "") && !g.name);
                        if (existing) existing.name = name;
                    }
                }
                matches.push({
                    index: match.index,
                    length: match[0].length,
                    value: match[0],
                    groups
                });
            }
        }

        return { success: true, matches, errorMessage: null };
    } catch (e) {
        return { success: false, matches: [], errorMessage: e.message };
    }
}

File: src/RequestKit.Shared/Interop/RegexInterop.cs
using System.Text.Json;
using Microsoft.JSInterop;
using RequestKit.Core.Models;

namespace RequestKit.Shared.Interop;

public class RegexInterop : IAsyncDisposable
{
    private readonly Lazy<Task<IJSObjectReference>> _moduleTask;

    public RegexInterop(IJSRuntime jsRuntime)
    {
        _moduleTask = new(() => jsRuntime.InvokeAsync<IJSObjectReference>(
            "import", "./_content/RequestKit.Shared/js/regexInterop.js").AsTask());
    }

    public async Task<(List<RegexMatch> Matches, string? Error)> ExecuteAsync(string pattern, string testString, string flags)
    {
        var module = await _moduleTask.Value;
        var json = await module.InvokeAsync<JsonElement>("execute", pattern, testString, flags);

        var success = json.GetProperty("success").GetBoolean();
        string? error = null;
        if (!success)
            error = json.GetProperty("errorMessage").GetString();

        var matches = new List<RegexMatch>();
        foreach (var m in json.GetProperty("matches").EnumerateArray())
        {
            var groups = new List<RegexGroup>();
            foreach (var g in m.GetProperty("groups").EnumerateArray())
            {
                groups.Add(new RegexGroup
                {
                    GroupIndex = g.GetProperty("groupIndex").GetInt32(),
                    Name = g.TryGetProperty("name", out var nameEl) && nameEl.ValueKind == JsonValueKind.String
                        ? nameEl.GetString()
                        : null,
                    Value = g.GetProperty("value").GetString() ?? "",
                    Index = g.GetProperty("index").GetInt32(),
                    Length = g.GetProperty("length").GetInt32()
                });
            }
            matches.Add(new RegexMatch
            {
                Index = m.GetProperty("index").GetInt32(),
                Length = m.GetProperty("length").GetInt32(),
                Value = m.GetProperty("value").GetString() ?? "",
                Groups = groups
            });
        }

        return (matches, error);
    }

    public async ValueTask DisposeAsync()
    {
        if (_moduleTask.IsValueCreated)
        {
            var module = await _moduleTask.Value;
            await module.DisposeAsync();
        }
    }
}
